File: app/src/main/java/com/kite/mnemoai/ui/mine/MineViewModel.java
package com.kite.mnemoai.ui.mine;

import static androidx.lifecycle.ViewModelProvider.AndroidViewModelFactory.APPLICATION_KEY;

import android.app.Application;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatDelegate;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MediatorLiveData;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.viewmodel.ViewModelInitializer;

import com.kite.mnemoai.MainApplication;
import com.kite.mnemoai.R;
import com.kite.mnemoai.data.repository.IRepositoryCallback;
import com.kite.mnemoai.data.repository.UserSettingRepository;
import com.kite.mnemoai.data.repository.WordRepository;
import com.kite.mnemoai.ui.mine.model.MineBaseItem;
import com.kite.mnemoai.ui.mine.model.MineSelectorItem;
import com.kite.mnemoai.ui.mine.model.MineTextItem;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

public class MineViewModel extends ViewModel {
    private MediatorLiveData<MineUIState> _uiStatus = new MediatorLiveData<>();
    private List<MineBaseItem> mineBaseItems = new ArrayList<>();
    private WordRepository wordRepository;
    private UserSettingRepository userSettingRepository;

    public MineViewModel(@NonNull Application app, WordRepository wordRepository, UserSettingRepository userSettingRepository) {
        this.wordRepository = wordRepository;
        this.userSettingRepository = userSettingRepository;
        _uiStatus.addSource(userSettingRepository.getUserSettingLiveData(), userSetting -> {
            List<MineBaseItem> newList = new ArrayList<>();
            MineSelectorItem themeSetting = new MineSelectorItem(app.getResources().getString(R.string.light_dark_model),
                    List.of(
                            app.getResources().getString(R.string.fallow_system),
                            app.getResources().getString(R.string.light_model),
                            app.getResources().getString(R.string.dark_model)
                    ),
                    convertLightDarkModelToOption(userSetting.getLightDarkModel()),
                    getThemeLastSelectedIndex(app.getResources().getString(R.string.light_dark_model), convertLightDarkModelToOption(userSetting.getLightDarkModel())),
                    (index) -> {
                        int lightDarkModel = userSetting.getLightDarkModel();
                        if (index == 0) {
                            if(lightDarkModel == -1) return;
                            AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_FOLLOW_SYSTEM);
                            saveLightDarkModel(-1);
                        } else if (index == 1) {
                            if(lightDarkModel == 1) return;
                            AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_NO);
                            saveLightDarkModel(1);
                        } else {
                            if(lightDarkModel == 2) return;
                            AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_YES);
                            saveLightDarkModel(2);
                        }
                    });
            newList.add(0, themeSetting);

            MineTextItem apiKeySetting = new MineTextItem(app.getResources().getString(R.string.setting_title_api_key),
                    userSetting.getApiKey());
            newList.add(1, apiKeySetting);

            mineBaseItems.clear();
            mineBaseItems.addAll(newList);
            updateUIStatus();
        });
    }

    private int convertLightDarkModelToOption(int lightDarkModel){
        if(lightDarkModel == AppCompatDelegate.MODE_NIGHT_FOLLOW_SYSTEM){
            return 0;
        }else if(lightDarkModel == AppCompatDelegate.MODE_NIGHT_NO){
            return 1;
        }else {
            return 2;
        }
    }

    private int getThemeLastSelectedIndex(String title, int lastSelectedIndex){
        for(MineBaseItem mineBaseItem: mineBaseItems){
            if(mineBaseItem instanceof MineSelectorItem){
                MineSelectorItem themeSetting = (MineSelectorItem) mineBaseItem;
                if(Objects.equals(themeSetting.getTitle(), title)){
                    return themeSetting.getSelected();
                }
            }
        }
        return lastSelectedIndex;
    }

    public void apiKeyTest(String apiKey, IRepositoryCallback<Boolean> callback){
        wordRepository.apiKeyValidTest(apiKey, callback);
    }

    public void saveApiKey(String apiKey){
        userSettingRepository.setApiKey(apiKey);
    }

    private void updateUIStatus(){
        _uiStatus.setValue(new MineUIState(mineBaseItems));
    }

    public LiveData<MineUIState> getUIStatus() {
        return _uiStatus;
    }

    private void saveLightDarkModel(int index){
        userSettingRepository.setLightDarkModel(index);
    }

    public final static ViewModelInitializer<MineViewModel> initializer = new ViewModelInitializer<MineViewModel>(
            MineViewModel.class,
            creationExtras -> {
                MainApplication app = (MainApplication) creationExtras.get(APPLICATION_KEY);
                assert app != null;
                return new MineViewModel(app, app.getWordRepository(), app.getUserSettingRepository());
            }
    );
}

File: app/src/main/java/com/kite/mnemoai/ui/mine/MineViewModel.kt
package com.kite.mnemoai.ui.mine

import android.app.Application
import androidx.appcompat.app.AppCompatDelegate
import androidx.lifecycle.LiveData
import androidx.lifecycle.MediatorLiveData
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.ViewModelProvider.AndroidViewModelFactory
import androidx.lifecycle.viewModelScope
import androidx.lifecycle.viewmodel.initializer
import androidx.lifecycle.viewmodel.viewModelFactory
import com.kite.mnemoai.MainApplication
import com.kite.mnemoai.R
import com.kite.mnemoai.data.local.UserSetting
import com.kite.mnemoai.data.network.WordExtractRequest
import com.kite.mnemoai.data.repository.AiMnemonicRepository
import com.kite.mnemoai.data.repository.UserSettingRepository
import com.kite.mnemoai.data.repository.WordRepository
import com.kite.mnemoai.ui.mine.model.MineBaseItem
import com.kite.mnemoai.ui.mine.model.MineSelectorItem
import com.kite.mnemoai.ui.mine.model.MineSelectorItem.OnOptionSelectedListener
import com.kite.mnemoai.ui.mine.model.MineTextItem
import com.kite.mnemoai.ui.model.LoadingState
import com.kite.mnemoai.ui.mine.model.ThemeType
import dagger.hilt.android.lifecycle.HiltViewModel
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.launch
import javax.inject.Inject

@HiltViewModel
class MineViewModel @Inject constructor(
    private val userSettingRepository: UserSettingRepository,
    private val aiMnemonicRepository: AiMnemonicRepository
) : ViewModel() {
    private val _uiStatus = MediatorLiveData<MineUIState?>()
    private var dayNightMode: Int = -1
    private var apiKey: String? = null
    private var apiTestState: LoadingState<String?>? = null

    init {
        _uiStatus.addSource(
            userSettingRepository.userSettingLiveData,
            Observer { userSetting: UserSetting ->
                dayNightMode = userSetting.lightDarkModel
                apiKey = userSetting.apiKey
                updateUIStatus()
            })
    }

    fun apiKeyTest(apiKey: String) {
        apiTestState = LoadingState.Loading
        updateUIStatus()
        viewModelScope.launch {
            apiTestState = aiMnemonicRepository.apiKeyTest(apiKey, WordExtractRequest("apple", false))
            if (apiTestState is LoadingState.Success<*>) {
                saveApiKey(apiKey)
            }
            updateUIStatus()
        }
    }

    fun saveApiKey(apiKey: String?) {
        userSettingRepository.setApiKey(apiKey)
    }

    private fun updateUIStatus() {
        _uiStatus.value = MineUIState(dayNightMode, apiKey, apiTestState)
    }

    val uIStatus: LiveData<MineUIState?>
        get() = _uiStatus

    fun saveLightDarkModel(lightDarkModel: Int) {
        dayNightMode = convertLightDarkModelToOption(lightDarkModel)
        userSettingRepository.setLightDarkModel(lightDarkModel)
        updateUIStatus()
    }

    fun convertLightDarkModelToOption(lightDarkModel: Int): Int {
        return when (lightDarkModel) {
            AppCompatDelegate.MODE_NIGHT_FOLLOW_SYSTEM -> {
                0
            }
            AppCompatDelegate.MODE_NIGHT_NO -> {
                1
            }
            else -> {
                2
            }
        }
    }
}
